### PR TITLE
chore: Add documentation for custom offline stores

### LINKED ISSUE #2411: Missing documentation for abstract `pull_all_from_table_or_query` offline store method
## Expected Behavior 

Since https://github.com/feast-dev/feast/pull/2197, offline store method `pull_all_from_table_or_query` must be overridden by custom offline stores. This is currently not documented.

Expectations: 
- [Docstring](https://github.com/feast-dev/feast/blob/b35e1e84720523cef70cba6d6306af8f193b469f/sdk/python/feast/infra/offline_stores/offline_store.py#L203) of `pull_all_from_table_or_query` contains a meaningful description of the method

- [Web doc](https://docs.feast.dev/how-to-guides/adding-a-new-offline-store) mentions that `pull_all_from_table_or_query` must be overriden.

## Current Behavior
No documentation for `pull_all_from_table_or_query`.

### PR DESCRIPTION
**Pull request by kevjumba**
Signed-off-by: Kevin Zhang <kzhang@tecton.ai>

**What this PR does / why we need it**:
Add some documentation on adding offline stores.
**Which issue(s) this PR fixes**:

Fixes #2411

### CODE REVIEW

File: docs/how-to-guides/adding-a-new-offline-store.md [hunk 1]
```diff
@@ -28,7 +28,8 @@ The OfflineStore class contains a couple of methods to read features from the of
 There are two methods that deal with reading data from the offline stores`get_historical_features`and `pull_latest_from_table_or_query`.
 
 * `pull_latest_from_table_or_query` is invoked when running materialization (using the `feast materialize` or `feast materialize-incremental` commands, or the corresponding `FeatureStore.materialize()` method. This method pull data from the offline store, and the `FeatureStore` class takes care of writing this data into the online store.
-* `get_historical_features `is invoked when reading values from the offline store using the `FeatureStore.get_historica_features()` method. Typically, this method is used to retrieve features when training ML models.
+* `get_historical_features `is invoked when reading values from the offline store using the `FeatureStore.get_historical_features()` method. Typically, this method is used to retrieve features when training ML models.
```
Comment (woop) [inline docs/how-to-guides/adding-a-new-offline-store.md:31]: typo

File: sdk/python/feast/infra/offline_stores/offline_store.py [hunk 2]
```diff
@@ -179,6 +179,19 @@ def pull_latest_from_table_or_query(
         end_date: datetime,
     ) -> RetrievalJob:
         """
+        This method is invoked when running materialization (using the feast materialize or feast
```
Comment (woop) [inline sdk/python/feast/infra/offline_stores/offline_store.py:182]: `Feast`, not `feast`

File: docs/how-to-guides/adding-a-new-offline-store.md (hunk 1 above)
Comment (kevjumba) [reply docs/how-to-guides/adding-a-new-offline-store.md:31]: Oh I think I fixed this before you commented.

File: sdk/python/feast/infra/offline_stores/offline_store.py (hunk 2 above)
Comment (kevjumba) [reply sdk/python/feast/infra/offline_stores/offline_store.py:182]: It is part of a command, added backticks to make more clear.

File: docs/how-to-guides/adding-a-new-offline-store.md [hunk 3]
```diff
@@ -66,6 +67,7 @@ There are two methods that deal with reading data from the offline stores`get_hi
                                                        created_timestamp_column,
                                                        start_date,
                                                        end_date)
+        )
```
Comment (felixwang9817) [inline docs/how-to-guides/adding-a-new-offline-store.md:70]: nit: I think this is unnecessary

File: sdk/python/feast/infra/offline_stores/offline_store.py [hunk 4]
```diff
@@ -179,6 +179,19 @@ def pull_latest_from_table_or_query(
         end_date: datetime,
     ) -> RetrievalJob:
         """
+        This method is invoked when running materialization (using the `feast materialize` or `feast
```
Comment (felixwang9817) [inline sdk/python/feast/infra/offline_stores/offline_store.py:182]: super nitty: can you put `This method pulls data from the offline store, and the FeatureStore class is used to write this data into the online store.` before `This method is invoked...`? in general I think a docstring should first describe what the function does, and only afterwards provide details on how it should be used, etc.

File: sdk/python/feast/infra/offline_stores/offline_store.py [hunk 5]
```diff
@@ -179,6 +179,19 @@ def pull_latest_from_table_or_query(
...
+        materialize-incremental` commands, or the corresponding FeatureStore.materialize() method.
+        This method pulls data from the offline store, and the FeatureStore class is used to write this
+        data into the online store.
+        Args:
```
Comment (felixwang9817) [inline sdk/python/feast/infra/offline_stores/offline_store.py:186]: nit: can this args section go at the end of the docstring? same with below

**Review by felixwang9817 — COMMENTED** [review]
just a couple of nits

**Review by felixwang9817 — APPROVED** [review]
/lgtm

**Pull request merged**